File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 */
#include <iostream>
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {
    ListNode *head = nullptr;
    ListNode *tail = nullptr;

    while (list1 != nullptr || list2 != nullptr) {
      ListNode *cur = new ListNode();
      if (list1 == nullptr) {
        cur->val = list2->val;
        list2 = list2->next;
      } else if (list2 == nullptr) {
        cur->val = list1->val;
        list1 = list1->next;
      } else {
        if (list1->val <= list2->val) {
          cur->val = list1->val;
          list1 = list1->next;
        } else {
          cur->val = list2->val;
          list2 = list2->next;
        }
      }

      if (head == nullptr)
        head = cur;
      if (tail == nullptr) {
        tail = cur;
      } else {
        tail->next = cur;
        tail = cur;
      }
    }
    return head;
  }
};

int main() {
  // Create two linked lists for testing
  ListNode *list1 = new ListNode(1);
  list1->next = new ListNode(3);
  list1->next->next = new ListNode(5);

  ListNode *list2 = new ListNode(2);
  list2->next = new ListNode(4);
  list2->next->next = new ListNode(6);

  Solution solution;

  // Merge the two lists
  ListNode *mergedList = solution.mergeTwoLists(list1, list2);

  // Print the merged list
  ListNode *cur = mergedList;
  while (cur != nullptr) {
    std::cout << cur->val << " ";
    cur = cur->next;
  }
  std::cout << std::endl;

  // Clean up the memory
  cur = mergedList;
  while (cur != nullptr) {
    ListNode *temp = cur;
    cur = cur->next;
    delete temp;
  }

  // Clean up the memory for the original lists
  cur = list1;
  while (cur != nullptr) {
    ListNode *temp = cur;
    cur = cur->next;
    delete temp;
  }

  cur = list2;
  while (cur != nullptr) {
    ListNode *temp = cur;
    cur = cur->next;
    delete temp;
  }

  return 0;
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        tail = dummy

        while list1 and list2:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            elif list2.val <= list1.val:
                tail.next = list2
                list2 = list2.next

            tail = tail.next

        if list1:
            tail.next = list1
        elif list2:
            tail.next = list2

        return dummy.next

def create_linked_list(values):
    if not values:
        return None
    
    head = ListNode(values[0])
    current = head
    for val in values[1:]:
        current.next = ListNode(val)
        current = current.next
    return head

def print_linked_list(head):
    current = head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")

def are_linked_lists_equal(head1, head2):
    current1, current2 = head1, head2
    while current1 and current2:
        if current1.val != current2.val:
            return False
        current1, current2 = current1.next, current2.next
    return current1 is None and current2 is None

values1 = [1, 2, 4]
values2 = [1, 3, 4]
expected_merged = [1, 1, 2, 3, 4, 4]

solution = Solution()

# Test case 1
list1 = create_linked_list(values1)
list2 = create_linked_list(values2)
merged = solution.mergeTwoLists(list1, list2)

print("Input List 1:", values1)
print("Input List 2:", values2)
print("Merged List:")
print_linked_list(merged)

if are_linked_lists_equal(merged, create_linked_list(expected_merged)):
    print("✅ Test passed")
else:
    print("❌ Test failed")

# Test case 2 (Empty Lists)
list1_empty = create_linked_list([])
list2_empty = create_linked_list([])
merged_empty = solution.mergeTwoLists(list1_empty, list2_empty)

print("\nInput List 1 (Empty):")
print_linked_list(list1_empty)
print("Input List 2 (Empty):")
print_linked_list(list2_empty)
print("Merged List (Empty):")
print_linked_list(merged_empty)

if are_linked_lists_equal(merged_empty, create_linked_list([])):
    print("✅ Test passed")
else:
    print("❌ Test failed")
